<a href="https://colab.research.google.com/github/emilioan55/TAREAS-MAESTRIA/blob/main/Actividad3BDManipulacion_A00042485.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: EMILIO ANDRES MENDOZA KAPLAN
*   MATRÍCULA: 00042485


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [1]:
pip install pymysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 1.6 MB/s eta 0:00:00


`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [2]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [3]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [4]:
# 1. Extrae toda la información de las películas que duran más de 5 horas.
query = sqla.text("""
    SELECT *
    FROM movie
    WHERE runtime > 300;
""")

df_long_movies = pd.read_sql(query, conn)
display(df_long_movies)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


**En la base de datos se localizaron cinco registros de películas con duración superior a cinco horas.**

2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [5]:
 #¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores NULL?
 #Cuenta los registros en los que el campo 'available_globally' tiene el valor FALSE y lo divide entre el total de registros para finalmente
 #multiplicar por 100 para dar el valor en porcentaje
query = sqla.text("""
    SELECT
        (SUM(CASE WHEN available_globally = false THEN 1 ELSE 0 END) * 100.0) / COUNT(*) AS porcentaje_solo_eeuu
    FROM movie
    WHERE available_globally IS NOT NULL;
""")

df_percentage = pd.read_sql(query, conn)
display(df_percentage)

,porcentaje_solo_eeuu
0,83.57175


**El 83.57 por ciento de las películas en la base de datos se encontraría disponible  unicamente en EEUU**

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [ ]:
# 3a. ¿Cuáles son los idiomas o regiones originales en la tabla de películas?
query_locales = sqla.text("""
    SELECT DISTINCT locale
    FROM movie
    WHERE locale IS NOT null;
""")

df_locales = pd.read_sql(query_locales, conn)
print("Idiomas o regiones originales presentes en la tabla:")
display(df_locales)

# 3b. ¿Cuántos registros tienen el campo locale con valor null?
query_nulls = sqla.text("""
    SELECT COUNT(*) AS total_null_locales
    FROM movie
    WHERE locale IS null;
""")

df_nulls = pd.read_sql(query_nulls, conn)
print("\nCantidad de registros con locale NULL:")
display(df_nulls)

Idiomas o regiones originales presentes en la tabla:


,locale
0,en



Cantidad de registros con locale NULL:


,total_null_locales
0,11255


**3.a La base de datos contiene solamente valores para las películas en inglés (en).**

**3.b 11,255 registros tienen como valor NULL en el campo 'locale' (idioma/region)**

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

4.1 Cuantas películas tienen títulos diferentes entre los campos "title" y "original_tile"

In [ ]:
#Determina cuántas películas tienen títulos diferentes los campos (title y original_title) cuando locale es NULL.
query_different_titles = sqla.text("""
    SELECT COUNT(*) AS total_diferentes
    FROM movie
    WHERE locale IS NULL
      AND title <> original_title;
""")

df_diff_titles = pd.read_sql(query_different_titles, conn)
display(df_diff_titles)

,total_diferentes
0,3947


4.2 ¿Coinciden los resultados (cantidad de NULL y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.

**No, no coinciden. De hecho el conteo de "null" en "locale" es de 11,255 registros, mientras que la de diferentes campos "title" y "original_title" es de 3,947. La característica de los registros faltantes está en que algunos títulos son iguales aunque no sean originarias de EEUU o en el campo de "original_title" el campo se relleno con "null"**

4.3 Finalmente, concluye si la suposición de que los valores NULL en locale indican que la película está en otro idioma es válida.

**No, no es necesariamente válida. Un ejemplo, de la observación de la base de datos, es:**
('2024-01-01 00:00:00.000', '2024-01-01 00:00:00.000', 'Despicable Me', null, 95, null, false, null),
**El útlimo campo ("locale" aparece como null y el campo "original_title" también es nulo**

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [ ]:
# Tenemos que buscar en la tabla view_summary en el campo "cummulative_weeks_in_top10"
# y encontrar el título de la película en la tabla movie a través del movie_id.
query_top_movie = sqla.text("""
    SELECT m.title, SUM(v.cumulative_weeks_in_top10) AS semanas_totales
    FROM view_summary v
    JOIN movie m ON v.movie_id = m.id
    GROUP BY m.id, m.title
    ORDER BY semanas_totales DESC
    LIMIT 1;
""")

df_top_movie = pd.read_sql(query_top_movie, conn)
print("\nPelícula que ha permanecido más tiempo en el top 10:")
display(df_top_movie)


Película que ha permanecido más tiempo en el top 10:


,title,semanas_totales
0,The Super Mario Bros. Movie,231.0


**Del análisis resulta que The Super Mario Bros.Movie ha permanecido más tiempo en el Top 10**

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [ ]:
#Cuenta las temporadas de cada programa, ordena de manera descencente y entrega los primros cinco
query_top_tv_shows = sqla.text("""
    SELECT t.title, COUNT(s.id) AS total_temporadas
    FROM tv_show t
    JOIN season s ON s.tv_show_id = t.id
    GROUP BY t.id, t.title
    ORDER BY total_temporadas DESC
    LIMIT 5;
""")

df_top_tv_shows = pd.read_sql(query_top_tv_shows, conn)
print("Los 5 programas de TV con mayor cantidad de temporadas:")
display(df_top_tv_shows)

Los 5 programas de TV con mayor cantidad de temporadas:


,title,total_temporadas
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


**Los cinco programas de TV con más temporadas en la base de datos son Grey's Anatomy (20), Naruto Shipupuden (20), Heartland (2007) (17), It's Always Sunny in Philadelphia (16) y NCIS (15)**

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [ ]:
# Encuentra para todos los campos de 'duration' en las que se especifica "SEMI_ANNUALLY" las fechas de inicio y de terminación
query_intervals = sqla.text("""
    SELECT duration, start_date, end_date, COUNT(*) AS total_registros
    FROM view_summary
    WHERE duration = 'SEMI_ANNUALLY'
    GROUP BY duration, start_date, end_date
    ORDER BY start_date;
""")

df_intervals = pd.read_sql(query_intervals, conn)
print("Intervalos de fechas para períodos semestrales (SEMI_ANNUALLY):")
display(df_intervals)

Object `semestrales` not found.
Intervalos de fechas para períodos semestrales (SEMI_ANNUALLY):


,duration,start_date,end_date,total_registros
0,SEMI_ANNUALLY,2023-07-01,2023-12-31,15763
1,SEMI_ANNUALLY,2024-01-01,2024-06-30,15912


**Los intervalos de fechas son dos, del 1 de agosto al 31 de diciembre de 2023 y del 1 de enero al 30 de junio de 2024**

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

Primero ordeno las temporadas del programa según la cantidad de vistas para el primer periodo semestral de 2024

In [ ]:
#Ordena las temporadas de Grey's Anatomy según la cantidad de vistas (primer período semestral de 2024).
query_greys = sqla.text("""
    SELECT s.title AS temporada, v.views AS vistas
    FROM tv_show t
    JOIN season s ON t.id = s.tv_show_id
    JOIN view_summary v ON s.id = v.season_id
    WHERE t.title = "Grey's Anatomy"
      AND v.duration = 'SEMI_ANNUALLY'
      AND v.start_date = '2024-01-01'
    ORDER BY v.views DESC;
""")

df_greys = pd.read_sql(query_greys, conn)
print("Temporadas de Grey's Anatomy ordenadas por vistas (1er semestre 2024):")
display(df_greys)


Temporadas de Grey's Anatomy ordenadas por vistas (1er semestre 2024):


,temporada,vistas
0,Grey's Anatomy: Season 1,3600000
1,Grey's Anatomy: Season 2,3100000
2,Grey's Anatomy: Season 3,2900000
3,Grey's Anatomy: Season 5,2900000
4,Grey's Anatomy: Season 4,2900000
5,Grey's Anatomy: Season 6,2800000
6,Grey's Anatomy: Season 7,2700000
7,Grey's Anatomy: Season 8,2600000
8,Grey's Anatomy: Season 9,2500000
9,Grey's Anatomy: Season 10,2400000


**La temporada 20 del programa Grey's Anatomy salió al aire el 14 de marzo de 2024, dentro del semestre del análisis.**

**Me parece notable que la cantidad de vistas en el primer semestre de 2024 (3,600,000) de la temporada 1, habiendo salido al aire el 27 de marzo de 2005, continuó atrayendo al público en 2024. En términos generales, las siguientes temporadas disminuyen en vistas en el semestre hasta poco menos de la mitad para las temporadas 16, 17 y 18.**

**La temporada 20 explicaría el número de vistas por su reciente salida al aire.**

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [ ]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [ ]:
# Cargar los dataframes necesarios desde la base de datos
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

# Filtrar el programa específico en tv_show
greys_tv = tv_show[tv_show['title'] == "Grey's Anatomy"]

# Unir tv_show filtrado con season (JOIN)
greys_season = pd.merge(greys_tv, season, left_on='id', right_on='tv_show_id', suffixes=('_tv', '_season'))

# Asegurarnos de que start_date sea manejable como string para filtrar más fácilmente
view_summary['start_date_str'] = view_summary['start_date'].astype(str)

# Filtrar view_summary por periodo y fecha (WHERE)
vs_filtered = view_summary[(view_summary['duration'] == 'SEMI_ANNUALLY') &
                           (view_summary['start_date_str'].str.startswith('2024-01-01'))]

# Unir las temporadas de Grey's Anatomy con el resumen de vistas filtrado
greys_views = pd.merge(greys_season, vs_filtered, left_on='id_season', right_on='season_id')

# Seleccionar columnas, renombrar y ordenar (SELECT y ORDER BY)
df_pandas = greys_views[['title_season', 'views']].rename(columns={'title_season': 'temporada', 'views': 'vistas'})
df_pandas = df_pandas.sort_values(by='vistas', ascending=False).reset_index(drop=True)

print("Temporadas de Grey's Anatomy ordenadas por vistas (Lógica de Pandas):")
display(df_pandas)

Temporadas de Grey's Anatomy ordenadas por vistas (Lógica de Pandas):


,temporada,vistas
0,Grey's Anatomy: Season 1,3600000
1,Grey's Anatomy: Season 2,3100000
2,Grey's Anatomy: Season 3,2900000
3,Grey's Anatomy: Season 5,2900000
4,Grey's Anatomy: Season 4,2900000
5,Grey's Anatomy: Season 6,2800000
6,Grey's Anatomy: Season 7,2700000
7,Grey's Anatomy: Season 8,2600000
8,Grey's Anatomy: Season 9,2500000
9,Grey's Anatomy: Season 10,2400000


**El resultado es el mismo, como se esperaría**

`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [ ]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Descargué el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y lo ubico en la carpeta donde lo almacené
import os

# Reemplaza esta ruta con la ubicación exacta de mi carpeta en Google Drive
DIR = "/content/drive/MyDrive/Colab Notebooks/Actividad 3 BDManipulacion/"

try:
    os.chdir(DIR)
    print(f"Directorio de trabajo cambiado exitosamente a:\n{os.getcwd()}")
except FileNotFoundError:
    print(f"Error: No se encontró la ruta '{DIR}'.")
    print("Por favor, verifica en el panel izquierdo (icono de carpeta) la ruta exacta donde guardaste el archivo 'consultancy.json' y actualiza la variable DIR.")

Directorio de trabajo cambiado exitosamente a:
/content/drive/MyDrive/Colab Notebooks/Actividad 3 BDManipulacion


In [ ]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que utilizo para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [ ]:
# Obtener los documentos de la colección EMPLOYEE
employees_ref = db.collection('EMPLOYEE')
docs = employees_ref.stream()

# Lista para guardar los datos extraídos
employees_data = []

for doc in docs:
    # doc.to_dict() obtiene los campos del documento
    data = doc.to_dict()
    # Agregamos manualmente el id del documento
    data['id'] = doc.id
    employees_data.append(data)

# Crear el DataFrame de Pandas
df_employees = pd.DataFrame(employees_data)

# Mostrar el dataframe resultante
print("Contenido de la colección EMPLOYEE:")
display(df_employees)

Contenido de la colección EMPLOYEE:


,emp_fname,emp_lname,id,emp_hiredate
0,David,Senior,8LcLuxVHGAd2d9IQc5jF,1989-07-12 06:00:00+00:00
1,William,Smithfield,Fzd60D6Z2CU4n0wVV8YN,2004-06-04 05:00:00+00:00
2,John,News,lX5xuQ5w3i6ib2ExccWY,2000-11-08 06:00:00+00:00
3,June,Arbough,yocFj2lichOkbAj9NBfp,1996-12-01 06:00:00+00:00


**El resultado se muestra en la tabla**

In [ ]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

*   OpenAI. (2026). *ChatGPT (basado en GPT-4)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://chat.openai.com/

---

**Utilicé Gemini Pro 3.1 como ayuda para la generación de algunos códigos.**